services/billing.py

```diff
@@ -155,18 +157,75 @@ def delete_subscription(self, owner: Owner):
             f"Downgrade to basic plan from user plan for owner {owner.ownerid} by user #{self.requesting_user.ownerid}",
             extra=dict(ownerid=owner.ownerid),
         )
+
         if subscription_schedule_id:
             log.info(
                 f"Releasing subscription from schedule for owner {owner.ownerid} by user #{self.requesting_user.ownerid}",
                 extra=dict(ownerid=owner.ownerid),
             )
             stripe.SubscriptionSchedule.release(subscription_schedule_id)
 
-        stripe.Subscription.modify(
-            owner.stripe_subscription_id,
-            cancel_at_period_end=True,
-            proration_behavior="none",
+        # we give an auto-refund grace period of 24 hours for a monthly subscription or 72 hours for a yearly subscription
+        current_subscription_datetime = datetime.fromtimestamp(
+            subscription["current_period_start"]
+        )
+        difference_from_now = datetime.now() - current_subscription_datetime
+
+        subscription_plan_interval = getattr(
+            getattr(subscription, "plan", None), "interval", None
         )
+        within_refund_grace_period = (
+            subscription_plan_interval == "month" and difference_from_now.days < 1
+        ) or (subscription_plan_interval == "year" and difference_from_now.days < 3)
+
+        customer = stripe.Customer.retrieve(owner.stripe_customer_id)
+        # we are giving customers 2 autorefund instances
+        autorefunds_remaining = int(
+            customer["metadata"].get("autorefunds_remaining", "2")
+        )
+
+        if within_refund_grace_period and autorefunds_remaining > 0:
+            stripe.Subscription.cancel(owner.stripe_subscription_id)
+
+            invoices_list = stripe.Invoice.list(
```

Comment:
it looks like you can add a created parameter to this call too and maybe shorten the amount of results to filter through as well. Not sure how much of an optimization... maybe to the point of not needing this stuff:

and invoice_created_datetime < current_subscription_datetime
and invoice_created_datetime >= start_of_last_period

Reply:
https://docs.stripe.com/api/invoices/list#list_invoices-created

Reply:
That would allow you to this variable above and condense this for loop as well

start_of_last_period = (
                    current_subscription_datetime - relativedelta(months=1)
                    if subscription_plan_interval == "month"
                    else current_subscription_datetime - relativedelta(years=1)
                )